### PR TITLE
feat(parser): canonicalize r. thymine input to u with W3020 soft-warning (closes #282)

### LINKED ISSUE #282: parser(rna): decide and pin r. thymine policy — reject `t` or canonicalize to `u` (#81 E follow-up to #232)
PR #232 (closes #231) pinned RNA lowercase enforcement coverage but left the `r.` thymine policy as a parser-policy decision: observed behavior is pinned only.

## Source

PR #232 § *Out of scope*:

> **`r.` thymine policy** — whether to reject `t` or canonicalize to `u` is a parser-policy decision, not part of E1. Observed behavior is pinned only.

## Decision required

Pick one of:

- **Reject** `r.123a>t` etc. at parse with a targeted diagnostic.
- **Canonicalize** to `u` (lenient) with a soft-validation warning.

Either path is a small parser change; this issue exists to capture the decision and the regression suite that should follow.

## Cross-ref

PR #232, #231 (closed). Fits #81 § E. Adjacent: #170 (preprocessor lowercase IUPAC) — orthogonal but in the same parser-policy neighborhood.

### PR DESCRIPTION
**Pull request by nh13**
## Summary

HGVS v21 specifies `r.` uses `u`. PR #232 pinned that ferro silently accepted `t` in lenient mode without surfacing the non-canonical input. PR #293 (issue #276) canonicalizes the Display side. This PR closes the loop on the parser side.

A new preprocessor phase `correct_rna_thymine` rewrites `t`/`T` → `u` in `r.` spans and emits the new **W3019 `RnaThymineCanonicalized`** soft-warning with the structured code. Strict mode rejects per standard soft-validation mode behaviour.

## Scoping

The phase tracks coord-prefix transitions to scope the rewrite to `r.` spans only. The boundary case `r.1a>t;c.5A>T` correctly leaves the `c.` half untouched. Unit tests pin the boundary at the `correct_rna_thymine` level so a future refactor can't silently widen the scope.

## Numbering

`PyErrorType::RnaThymineCanonicalized = 30` to leave room for the in-flight W3016–W3018 codes claimed by PRs #295, #272, #298. A comment in `python.rs` documents the skip so future authors don't reuse 27/28/29.

## Tests

`tests/issue_282_rna_thymine_policy.rs` — 30 tests across 10 modules (substitution, deletion, insertion, duplication, inversion, delins, repeat, compound brackets, mode behavior, non-RNA guards). Plus 3 unit tests in `corrections.rs` pinning the RNA→non-RNA boundary directly.

## Review

Two-reviewer cycle (Opus + Sonnet). Both verdicts: LGTM. Feedback (CHANGELOG, doc-vs-code mismatch, boundary test, registry uppercase tweak, PyErrorType reservation comment) addressed in this commit.

## Note on commit signing

1Password biometric failed three times; commit is `--no-gpg-sign` per project rules.

## Test plan

- [x] `cargo nextest run --features dev` — 4795 pass / 0 fail / 51 skipped.
- [x] `cargo check --features python` — clean.
- [x] `cargo clippy --features dev --all-targets -- -D warnings` — clean.
- [x] `cargo fmt --all` — clean.

Closes #282. Follow-up to PR #232 / #231 (#81 § E1). Sibling to PR #293 (#276) on the Display side.

### CODE REVIEW

**Comment by nh13** [comment]
Hourly maintenance bot: attempted `git rebase origin/main` and got 5 conflicts (`docs/spec/error_code_audit.md`, `src/error_handling/registry.rs`, `src/error_handling/types.rs`, `src/python.rs`, `tests/fixtures/error_code_audit.json`). These touch the error-code registry where a structural merge needs spec-aware judgement; aborted rather than risk an inconsistent W-code state. Needs human rebase. Last attempt: 2026-05-19T10:41:56Z.

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

File: docs/spec/error_code_audit.md [hunk 1]
```diff
@@ -104,6 +104,7 @@ and the soft-validation issues
...
+| W3020 | RnaThymineCanonicalized — `t`/`T` inside an `r.` (RNA) description rewritten to `u` | [recommendations/RNA/substitution.md](https://hgvs-nomenclature.org/stable/recommendations/RNA/substitution/) and adjacent RNA edit pages (RNA alphabet is `a/c/g/u`; `t` is non-canonical) | Enforced — pre-parse rewrite in lenient/silent modes via `correct_rna_thymine` (preprocessor Phase 16a); strict rejects with `E1006 UnexpectedChar` + hint. Closes the input-side loop on issue #232's deferred policy decision (closes #282); PR #276 / Display canonicalization handles the output side independently for raw `parse_hgvs` callers |
```
[the remark this answers is not in the document]
Comment (nh13) [reply docs/spec/error_code_audit.md]: Updated both the audit table (`docs/spec/error_code_audit.md` line 107) and its fixture mirror (`tests/fixtures/error_code_audit.json` W3020 `spec_section_notes`) to reference `PR #293` (display-side r. T→u canonicalization PR) instead of `PR #276` (the underlying issue). All 7 `error_code_audit` tests pass. Addressed in eccac1ebef492613f74dbb1385b70c5ee1bc8912.

**Comment by nh13** [comment]
@coderabbitai review

File: src/error_handling/corrections.rs [hunk 2]
```diff
@@ -2742,6 +2742,70 @@ pub fn correct_redundant_repeat_label(input: &str) -> (String, Vec<DetectedCorre
...
+                i,
+                i + 1,
+            ));
+            result.push('u');
+            i += 1;
+            continue;
+        }
+        let ch_end = next_char_end(bytes, i);
+        result.push_str(&input[i..ch_end]);
+        i = ch_end;
```
[the remark this answers is not in the document]
Comment (nh13) [reply src/error_handling/corrections.rs:2960]: Confirmed the bug — `r.1a>t;NT_000001.1:c.5A>T` corrupts `NT_` to `Nu_` under the old code (the existing test happened to use `NM_` and missed it because `M` isn't in the rewrite set).

Fix: added bracket-depth tracking so a top-level (depth 0) `;` or `|` ends the in_rna span, while `;`/`|` inside a bracketed compound (`r.[a;b]`) preserves the span as before. Updated the doc-comment scoping rules to reflect the new boundary.

Pinned 3 new tests:
- `test_correct_rna_thymine_top_level_semicolon_ends_span` — locks in the CR case (`NT_` unchanged)
- `test_correct_rna_thymine_top_level_pipe_ends_span` — same for `|`
- `test_correct_rna_thymine_bracketed_compound_keeps_span` — locks in that `r.[1a>t;2t>g]` still rewrites both halves

Full suite 5156/5156 passes, clippy clean. Addressed in 11d179b4b2fd8113a1a7297ecf541c9b515f22b4.

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review
